src/databricks/labs/ucx/source_code/redash.py

```diff
@@ -18,41 +18,44 @@
 class Redash:
     MIGRATED_TAG = "Migrated by UCX"
 
-    def __init__(self, index: TableMigrationIndex, ws: WorkspaceClient, installation: Installation):
+    def __init__(
+        self,
+        index: TableMigrationIndex,
+        ws: WorkspaceClient,
+        installation: Installation,
+        dashboard_crawler: RedashDashboardCrawler,
+    ):
         self._index = index
         self._ws = ws
         self._installation = installation
+        self._crawler = dashboard_crawler
 
-    def migrate_dashboards(self, dashboard_id: str | None = None) -> None:
-        for dashboard in self._list_dashboards(dashboard_id):
-            assert dashboard.id is not None
-            if dashboard.tags is not None and self.MIGRATED_TAG in dashboard.tags:
+    def migrate_dashboards(self, *dashboard_ids: str) -> None:
+        for dashboard in self._list_dashboards(*dashboard_ids):
+            if self.MIGRATED_TAG in dashboard.tags:
                 logger.debug(f"Dashboard {dashboard.name} already migrated by UCX")
                 continue
-            for query in self.get_queries_from_dashboard(dashboard):
+            for query in self._crawler.list_queries(dashboard):
                 self._fix_query(query)
             self._ws.dashboards.update(dashboard.id, tags=self._get_migrated_tags(dashboard.tags))
 
-    def revert_dashboards(self, dashboard_id: str | None = None) -> None:
-        for dashboard in self._list_dashboards(dashboard_id):
-            assert dashboard.id is not None
-            if dashboard.tags is None or self.MIGRATED_TAG not in dashboard.tags:
+    def revert_dashboards(self, *dashboard_ids: str) -> None:
+        for dashboard in self._list_dashboards(*dashboard_ids, force_refresh=True):  # Refresh for up-to-date tags
+            if self.MIGRATED_TAG not in dashboard.tags:
                 logger.debug(f"Dashboard {dashboard.name} was not migrated by UCX")
                 continue
-            for query in self.get_queries_from_dashboard(dashboard):
+            for query in self._crawler.list_queries(dashboard):
                 self._revert_query(query)
             self._ws.dashboards.update(dashboard.id, tags=self._get_original_tags(dashboard.tags))
 
-    def _list_dashboards(self, dashboard_id: str | None) -> list[Dashboard]:
-        try:
-            if dashboard_id is None:
-                return list(self._ws.dashboards.list())
-            return [self._ws.dashboards.get(dashboard_id)]
-        except DatabricksError as e:
-            logger.warning(f"Cannot list dashboards: {e}")
-            return []
+    def _list_dashboards(self, *dashboard_ids: str, force_refresh: bool = False) -> list[Dashboard]:
```

Comment:
We shouldn't allow forcing a refresh here: that's supposed to be limited to just workflow tasks (directly) because we want it to be clear where refreshes are happening.

If the caller expects a refresh, it needs to be explicitly done by the workflow beforehand.

Reply:
This is difficult as the revert dashboard requires the latest tags (to include the migration tag) and the dashboard migration is a cli command, not a workflow.

To resolve your comment, the dashboard migration needs to be transpiled to a workflow, which I do not want to include in the scope of this PR

Reply:
If not the workflow, then the CLI should do this: the intent is that explicit refreshing is supposed to be a 'top-level' action and not occur implicitly during other activity.